labelbox/orm/model.py

```diff
@@ -239,11 +239,58 @@ class EntityMeta(type):
     of the Entity class object so they can be referenced for example like:
         Entity.Project.
     """
+    mappings: Dict[str, "Entity"] = {}
 
     def __init__(cls, clsname, superclasses, attributedict):
         super().__init__(clsname, superclasses, attributedict)
+        cls.validate_cached_relationships()
         if clsname != "Entity":
             setattr(Entity, clsname, cls)
+            EntityMeta.mappings[utils.snake_case(
+                cls.__name__)] = cls.relationships()
+
+    @staticmethod
+    def raise_for_nested_cache(first: str, middle: str, last: str):
+        raise TypeError(
+            "Cannot cache a relationship to an Entity with its own cached relationship(s). "
+            f"`{first}` caches `{middle}` which caches `{last}`")
+
+    def validate_cached_relationships(cls):
+        """
+        Graphql doesn't allow for infinite nesting in queries. 
+        This function checks that cached relationships result in valid queries.
+            * It does this by making sure that a cached relationship do not
+              reference any entity with its own cached relationships.
+
+        This check is performed by looking to see if this entity caches 
+        any entities that have their own cached fields. If this entity 
+        that we are checking has any cached fields then we also check 
+        all currently defined entities to see if they cache this entity.
+
+        A two way check is necessary because checks are performed as classes are being defined.
+        As opposed to after all objects have been created.
```

Comment:
I'll continue reviewing tomorrow, but at a first glance: isn't this code equivalent to detecting a cycle in relationship graph? If so, doing a two-way check doesn't seem enough?

Reply:
This should be sufficient because the relationship that we are checking is only between two Entities and there are only three conditions.

1. An object is defined that doesn't have any cache.
2. A parent is defined and caches a child that isn't defined
    * Later that child is defined and checks for if this parent caches it and it has its own cached children. If so, raise else pass.
3. A child is defined that caches an object.
    * Later a parent is defined and checks if this child caches another object. If so, raise else pass.